### PR TITLE
Mqtt5Client Clean shutdown

### PR DESCRIPTION
**Pull request by xiazhvera**
*Issue #, if available:*

*Description of changes:*
- Added the self-reference share_ptr to Mqtt5Client to avoid the client get released before underlying native client clean up the operations and terminated.
- Updated the callbacks to use raws client instead of shared_ptr
- Added a Close() API to handle client shutdown.
- Improve NewMqtt5Client, so it will return nullptr when client creation failed. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: bin/mqtt5_app/main.cpp [hunk 1]
```diff
@@ -581,5 +583,6 @@ int main(int argc, char **argv)
     {
         std::cout << "[ERROR]Failed to start the client " << std::endl;
     }
+    mqtt5Client->Close();
```
Comment (sbSteveK) [inline bin/mqtt5_app/main.cpp:586]: Maybe we can add a comment that this must be called to clean up for customers.

File: include/aws/crt/mqtt/Mqtt5Client.h [hunk 2]
```diff
@@ -316,6 +310,20 @@ namespace Aws
                     std::shared_ptr<UnsubscribePacket> unsubscribeOptions,
                     OnUnsubscribeCompletionHandler onUnsubscribeCompletionCallback = NULL) noexcept;
 
+                /*
+                 * Tell the client to release the native client and clean up unhandled the resources
```
Comment (sbSteveK) [inline include/aws/crt/mqtt/Mqtt5Client.h:314]: maybe a typo here?

File: source/mqtt/Mqtt5Client.cpp [hunk 3]
```diff
@@ -362,25 +365,43 @@ namespace Aws
...
+                    m_client = aws_mqtt5_client_new(allocator, &clientOptions);
+                }
             }
 
             Mqtt5Client::~Mqtt5Client()
             {
+                aws_rw_lock_wlock(&m_client_lock);
                 if (m_client != nullptr)
                 {
+                    // Techniquely, this should never happen as the Mqtt5Client will not get released until we call
```
Comment (sbSteveK) [inline source/mqtt/Mqtt5Client.cpp:379]: trivial: same as above /* */ for consistency?
Technically* typo

File: include/aws/crt/mqtt/Mqtt5Client.h (hunk 2 above)
Comment (TwistedTwigleg) [reply include/aws/crt/mqtt/Mqtt5Client.h:314]: Minor: `/*` -> `/**`

File: source/mqtt/Mqtt5Client.cpp [hunk 4]
```diff
@@ -171,46 +173,45 @@ namespace Aws
...
-                        case aws_mqtt5_packet_type::AWS_MQTT5_PT_NONE:
-                        {
-                            publish = std::make_shared<PublishResult>(error_code);
-                            break;
+                            std::shared_ptr<PubAckPacket> packet = std::make_shared<PubAckPacket>(
+                                *(aws_mqtt5_packet_puback_view *)publshCompletionPacket, callbackData->allocator);
+                            publish = std::make_shared<PublishResult>(std::move(packet));
                         }
-                        default: // Invalid packet type
+                        else // This should never happened.
```
Comment (TwistedTwigleg) [inline source/mqtt/Mqtt5Client.cpp:190]: Minor: `//` -> `/* ... */`

File: source/mqtt/Mqtt5Client.cpp [hunk 5]
```diff
@@ -171,46 +173,45 @@ namespace Aws
...
-                    if (callbackData->onPublishCompletion != NULL)
+                    case aws_mqtt5_packet_type::AWS_MQTT5_PT_NONE:
                     {
-                        callbackData->onPublishCompletion(callbackData->client, error_code, publish);
+                        publish = std::make_shared<PublishResult>(error_code);
+                        break;
                     }
-
-                    Crt::Delete(callbackData, callbackData->allocator);
+                    default: // Invalid packet type
```
Comment (TwistedTwigleg) [inline source/mqtt/Mqtt5Client.cpp:202]: Minor: `//` -> `/* ... */`

File: source/mqtt/Mqtt5Client.cpp [hunk 6]
```diff
@@ -362,25 +365,43 @@ namespace Aws
...
+                }
             }
 
             Mqtt5Client::~Mqtt5Client()
             {
+                aws_rw_lock_wlock(&m_client_lock);
                 if (m_client != nullptr)
                 {
+                    // Techniquely, this should never happen as the Mqtt5Client will not get released until we call
+                    // Close(), and in Close() we should have released the native client already. WARN THE USER THERE IS
```
Comment (TwistedTwigleg) [inline source/mqtt/Mqtt5Client.cpp:380]: Minor: `Close()` is missing "`"

File: source/mqtt/Mqtt5Client.cpp [hunk 7]
```diff
@@ -362,25 +365,43 @@ namespace Aws
...
+                    // SOMETHING WRONG HERE.
+                    AWS_LOGF_WARN(
```
Comment (TwistedTwigleg) [inline source/mqtt/Mqtt5Client.cpp:382]: Debatable: I'd make this an ERROR

File: source/mqtt/Mqtt5Client.cpp [hunk 8]
```diff
@@ -389,20 +410,59 @@ namespace Aws
                 }
 
                 toSeat = new (toSeat) Mqtt5Client(options, allocator);
-                return std::shared_ptr<Mqtt5Client>(
+                // Creation failed, make sure we release the allocated memory
```
Comment (TwistedTwigleg) [inline source/mqtt/Mqtt5Client.cpp:413]: Minor: `//` -> `/* ... */`

File: source/mqtt/Mqtt5Client.cpp [hunk 9]
```diff
@@ -389,20 +410,59 @@ namespace Aws
...
             int Mqtt5Client::LastError() const noexcept { return aws_last_error(); }
 
-            bool Mqtt5Client::Start() const noexcept { return aws_mqtt5_client_start(m_client) == AWS_OP_SUCCESS; }
+            bool Mqtt5Client::Start() noexcept
+            {
+                if (aws_rw_lock_try_rlock(&m_client_lock) != AWS_ERROR_SUCCESS || m_client == nullptr)
+                {
+                    AWS_LOGF_ERROR(
+                        AWS_LS_MQTT5_CLIENT,
+                        "Mqtt5Client: Invalid Client or the client is in termination. Please create a new client.");
```
Comment (TwistedTwigleg) [inline source/mqtt/Mqtt5Client.cpp:436]: Minor: I'd either change `Invalid Client ...` to `Invalid client` or change `... the client is ...` to `... the Client is ...` for consistency. Same change will be needed for `stop()`, `publish()`, etc.

File: source/mqtt/Mqtt5Client.cpp [hunk 10]
```diff
@@ -516,19 +600,40 @@ namespace Aws
                 return result == AWS_OP_SUCCESS;
             }
 
+            void Mqtt5Client::Close() noexcept
+            {
+                aws_rw_lock_wlock(&m_client_lock);
+                if (m_client != nullptr)
+                {
+                    aws_mqtt5_client_release(m_client);
+                    // Waiting for client termination
```
Comment (TwistedTwigleg) [inline source/mqtt/Mqtt5Client.cpp:609]: Minor: `//` -> `/* ... */`

File: source/mqtt/Mqtt5Client.cpp [hunk 11]
```diff
@@ -171,46 +173,45 @@ namespace Aws
...
-                        case aws_mqtt5_packet_type::AWS_MQTT5_PT_NONE:
-                        {
-                            publish = std::make_shared<PublishResult>(error_code);
-                            break;
+                            std::shared_ptr<PubAckPacket> packet = std::make_shared<PubAckPacket>(
+                                *(aws_mqtt5_packet_puback_view *)publshCompletionPacket, callbackData->allocator);
+                            publish = std::make_shared<PublishResult>(std::move(packet));
                         }
-                        default: // Invalid packet type
+                        else /* This should never happened. */
```
Comment (bretambrose) [inline source/mqtt/Mqtt5Client.cpp:190]: I'd make it a fatal assert.

File: source/mqtt/Mqtt5Client.cpp [hunk 12]
```diff
@@ -171,46 +173,45 @@ namespace Aws
...
-                        callbackData->onPublishCompletion(callbackData->client, error_code, publish);
+                        publish = std::make_shared<PublishResult>(error_code);
+                        break;
                     }
-
-                    Crt::Delete(callbackData, callbackData->allocator);
+                    default: /* Invalid packet type */
+                    {
+                        AWS_LOGF_INFO(AWS_LS_MQTT5_CLIENT, "Invalid Packet Type.");
+                        publish = std::make_shared<PublishResult>(AWS_ERROR_INVALID_ARGUMENT);
```
Comment (bretambrose) [inline source/mqtt/Mqtt5Client.cpp:205]: Invalid argument is intended for user-submitted data, not crt-generated data.  I would use a different error code here and above.

File: source/mqtt/Mqtt5Client.cpp [hunk 13]
```diff
@@ -362,25 +365,47 @@ namespace Aws
                 clientOptions.client_termination_handler = &Mqtt5Client::s_clientTerminationCompletion;
                 clientOptions.client_termination_handler_user_data = this;
 
-                m_client = aws_mqtt5_client_new(allocator, &clientOptions);
+                if (aws_rw_lock_init(&m_client_lock) == AWS_OP_SUCCESS)
```
Comment (bretambrose) [inline source/mqtt/Mqtt5Client.cpp:368]: I would just fatal assert this.

File: source/mqtt/Mqtt5Client.cpp [hunk 14]
```diff
@@ -362,25 +365,47 @@ namespace Aws
...
+                    m_client = aws_mqtt5_client_new(allocator, &clientOptions);
+                }
             }
 
             Mqtt5Client::~Mqtt5Client()
             {
+                aws_rw_lock_wlock(&m_client_lock);
                 if (m_client != nullptr)
                 {
+                    /**
```
Comment (bretambrose) [inline source/mqtt/Mqtt5Client.cpp:379]: Really skeptical that any of this (if-block) is necessary or useful.

File: source/mqtt/Mqtt5Client.cpp [hunk 15]
```diff
@@ -389,20 +414,59 @@ namespace Aws
...
             }
 
             Mqtt5Client::operator bool() const noexcept { return m_client != nullptr; }
 
             int Mqtt5Client::LastError() const noexcept { return aws_last_error(); }
 
-            bool Mqtt5Client::Start() const noexcept { return aws_mqtt5_client_start(m_client) == AWS_OP_SUCCESS; }
+            bool Mqtt5Client::Start() noexcept
+            {
+                if (aws_rw_lock_try_rlock(&m_client_lock) != AWS_ERROR_SUCCESS || m_client == nullptr)
```
Comment (bretambrose) [inline source/mqtt/Mqtt5Client.cpp:436]: This pattern doesn't unlock a taken read lock if the client is null.

Maybe we should create a simple scoped rw lock class where creation fatal-assert-takes the lock and destruction releases it.

Also you can't release the lock until the C function returns.  Close() could have decrefed it inbetween unlock and aws_mqtt5_* invocation.

These apply to all usages of this pattern.

File: source/mqtt/Mqtt5Client.cpp [hunk 16]
```diff
@@ -516,19 +604,40 @@ namespace Aws
...
 
+            void Mqtt5Client::Close() noexcept
+            {
+                aws_rw_lock_wlock(&m_client_lock);
+                if (m_client != nullptr)
+                {
+                    aws_mqtt5_client_release(m_client);
+                    /* Waiting for client termination */
+                    std::unique_lock<std::mutex> lock(m_terminationMutex);
+                    m_terminationCondition.wait(lock, [this] { return m_terminationPredicate == true; });
```
Comment (bretambrose) [inline source/mqtt/Mqtt5Client.cpp:615]: This is a deadlock if someone calls Close() from an event loop triggered callback.

Why do we need to wait anyways?

File: source/mqtt/Mqtt5Client.cpp (hunk 4 above)
Comment (bretambrose) [reply source/mqtt/Mqtt5Client.cpp]: // comments are fine in C++

File: source/mqtt/Mqtt5Client.cpp (hunk 12 above)
Comment (xiazhvera) [reply source/mqtt/Mqtt5Client.cpp]: I checked out the error codes, but I did not find anything suitable for the case here, I used AWS_ERROR_UNKNOWN for now.

**Comment by xiazhvera** [comment]
New solution proposed. Close the PR.

**Pull request closed without merging**